src/main/java/com/qubit/solution/fenixedu/integration/cgd/services/utils/MoreThanOneAssociatedUnitStrategy.java

```diff
@@ -0,0 +1,79 @@
+package com.qubit.solution.fenixedu.integration.cgd.services.utils;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Set;
+
+import org.fenixedu.academic.domain.organizationalStructure.Unit;
+import org.fenixedu.academic.domain.student.Registration;
+
+import com.qubit.solution.fenixedu.integration.cgd.domain.configuration.CgdIntegrationConfiguration;
+
+public class MoreThanOneAssociatedUnitStrategy extends BaseIESCodeProviderStrategy implements CgdIESCodeProviderStrategyClass {
+
+    @Override
+    public List<String> getIESCode(Registration registration) {
+        Unit associatedUnit = registration.getDegree().getUnit();
+
+        if (associatedUnit == null) {
+            return super.getIESCode(registration);
+        }
+
+        return CgdIntegrationConfiguration.getInstance().getAllowsMultipleUnits() ? findAllUnitsCodes(registration,
+                associatedUnit) : findFirstUnitFromConfiguration(registration, associatedUnit);
+    }
+
+    private List<String> findAllUnitsCodes(Registration registration, Unit associatedUnit) {
+        Set<Unit> allowedUnits = CgdIntegrationConfiguration.getInstance().getUnitsSet();
+        List<String> result = new ArrayList<String>();
+
+        if (allowedUnits.contains(associatedUnit)) {
+            String code = associatedUnit.getCode();
+            if (!result.contains(code)) {
```

Comment:
@miguel-morais if result is a Set<String> you don't need to worry about checking if it exists and you just add it.